aerogear/data-sync-server
# Metrics endpoint + resolver metrics

https://issues.jboss.org/browse/AEROGEAR-7650

### Checklist
- [x] metrics endpoint with basic metrics
- [x] resolver timing metrics (time period between request & response of database queries)
- [x] collecting number of queries/mutations made (possibly split to un/successful ones)
- [x] change the way that request information is passed to metrics
- [x] "inmemory" type resolver metrics
- [x] logging of dataSourceType (in logger & metrics)
- [x] server response metrics

### How to verify this
`npm install` (new dependency **prom-client** has been added)
`docker-compose -p aerogeardatasyncserver up`
`npm run db:init:memeo:postgres`
`npm run dev`

1. Go to http://localhost:8000/metrics and check that metrics data is available
1. Go to http://localhost:8000/graphiql, paste in following and run it:
```
mutation createprofile {
  createProfile(email: "jon@doh.com", displayname: "Jon D Oh", pictureurl: "https://cdn1.thr.com/sites/default/files/imagecache/scale_crop_768_433/2011/10/doh_a.jpg") {
    email
    id
    displayname
  }
}

query getAllMemes {
  allMemes {
    id
    likes
    comments {
      id
      owner
      comment
    }
  }
}
```
3. Refresh http://localhost:8000/metrics and check that you see a record like 
```
resolver_timing_ms{datasource_type="Postgres",operation_type="mutation",name="createProfile"} 5
resolver_timing_ms{datasource_type="Postgres",operation_type="query",name="allMemes"} 3

requests_resolved{datasource_type="Postgres",operation_type="query",path="Query.allMemes"} 1
requests_resolved{datasource_type="Postgres",operation_type="mutation",path="Mutation.createProfile"}
```

4. There should be also record about GET (`/metrics`) and POST (`/graphiql?<query>`) requests, like:
```
server_response_ms{request_type="POST",error="false"} 9
server_response_ms{request_type="GET",error="false"} 8
```
5. When you run other queries, those values should change
